src/js/core/polyfills.js

```diff
@@ -102,11 +102,42 @@ function domPolyfills() {
     })([Element.prototype, CharacterData.prototype, DocumentType.prototype]);
 }
 
+function JSXPolyfills() {
+    // https://betterprogramming.pub/how-to-use-jsx-without-react-21d23346e5dc
+    const appendChild = (parent, child) => {
+        if (Array.isArray(child)) child.forEach(nestedChild => appendChild(parent, nestedChild));
+        else parent.appendChild(child.nodeType ? child : document.createTextNode(child));
+    };
+
+    window.JSXCreateElement = (tag, props, ...children) => {
+        if (typeof tag === "function") return tag(props, children);
+
+        const element = document.createElement(tag);
+
+        Object.entries(props || {}).forEach(([name, value]) => {
+            if (name.startsWith("on") && name.toLowerCase() in window)
+                element.addEventListener(name.toLowerCase().substring(2), value);
+            else element.setAttribute(name, value.toString());
+        });
+
+        children.forEach(child => {
+            appendChild(element, child);
+        });
+
+        return element;
+    };
+
+    window.JSXCreateFragment = (props, ...children) => {
+        return children;
+    };
```

Comment:
I feel like flattening fragments immediately instead of when actually creating the HTML element would make more sense, since it means intermediate fragment values are normalized

Reply:
switched sides; it's kinda inefficient to re-flatten sub-elements every time they're nested, and it really doesn't matter besides  for debugging